### DISCUSSION TITLE
0347. 前 K 个高频元素

### DISCUSSION THREAD

**Discussion by giscus** [itcharge/AlgoNote#118] [Comments]
# 0347. 前 K 个高频元素

[https://algo.itcharge.cn/solutions/0300-0399/top-k-frequent-elements/](https://algo.itcharge.cn/solutions/0300-0399/top-k-frequent-elements/)

**Comment by s1acr** [comment]
用heapq模块的解法
```py
import heapq
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # 优先队列
        nums_dict = dict()
        for num in nums:
            if num not in nums_dict:
                nums_dict[num] = 1
            else:
                nums_dict[num] += 1

        # 降重
        nums_list = list(set(nums))
        # or nums_dict.keys()

        q = []
        for i in range(len(nums_list)):
            heapq.heappush(q, (nums_dict[nums_list[i]], nums_list[i]))
        return [item[1] for item in heapq.nlargest(k, q)][:k]
   ```

**Reply by s1acr** [reply]
最后那里写错了, 没有[:k]

**Comment by WangDHXhao** [comment]
可以先计算频次，然后再计算前k个高频的数
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        nums_dict={}
        for num in nums:
            if num in nums_dict:
                nums_dict[num]+=1
            else:
                nums_dict[num]=1
        sorted_nums_dict=dict(sorted(nums_dict.items(),key=lambda item:item[1],reverse=True))
        result=list(sorted_nums_dict.keys())[:k]  
        return result
